### PR TITLE
Cherrypick #95981 to 1.18, Enables HTTP/2 health check

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Backports use of the golang.org/x/net PingTimeout to 1.18 (replay of #96778 on 1.18)

Rationale for backporting:

* how unpredictable the stuck connection issue is (can be triggered by arbitrary network interruptions)
* how many components it affects (all components using client-go)
* how problematic a stuck connection is for informer clients (can cause permanently stale data with no indication anything is wrong)
* we've successfully used these sys and net levels in 1.19 for O(months) now
* other consumers have successfully backported these changes to 1.18 and used them in production

#### Does this PR introduce a user-facing change?
```release-note
HTTP/2 connection health check is enabled by default in all Kubernetes clients to fix persistently broken connections (https://github.com/kubernetes/client-go/issues/374). If needed, users can tune the feature via the HTTP2_READ_IDLE_TIMEOUT_SECONDS and HTTP2_PING_TIMEOUT_SECONDS environment variables. The feature is disabled if HTTP2_READ_IDLE_TIMEOUT_SECONDS is set to 0.
```

/cc @caesarxuchao

### CODE REVIEW

**Comment by caesarxuchao** [comment]
/assign

Thanks, Jordan!

File: staging/src/k8s.io/client-go/rest/connection_test.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+)
+
+func TestReconnectBrokenTCP(t *testing.T) {
+	defer setEnv("HTTP2_READ_IDLE_TIMEOUT_SECONDS", strconv.Itoa(readIdleTimeout))()
+	defer setEnv("HTTP2_PING_TIMEOUT_SECONDS", strconv.Itoa(pingTimeout))()
+	defer setEnv("DISABLE_HTTP2", "")()
+	ts := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "Hello, %s", r.Proto)
+	}))
+	ts.EnableHTTP2 = true
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/connection_test.go:101]: this is new in go1.14, need to adjust the test for go1.13

Comment (liggitt) [reply staging/src/k8s.io/client-go/rest/connection_test.go]: fixed

**Comment by liggitt** [comment]
failure looks like #98182 

/retest

**Comment by liggitt** [comment]
conceptually acked by @deads2k in https://github.com/kubernetes/kubernetes/pull/95981#issuecomment-806141558, leaning on @caesarxuchao for review

**Comment by LuChenjing** [comment]
@liggitt Hi，I still met this issue after upgrading K8S from 1.18.5 to 1.18.20. I noticed this PR was release after 1.18.18.
How I upgraded K8S is replacing the apiserver\scheduler\controller manager\kubelet\kube-proxy docker images, is there any other configs I should set?

**Comment by xuchen-xiaoying** [comment]
it seems that after this cherry pick, it still reproduce problem that "use of closed network connection". Should I also update go version ?

**Comment by xuchen-xiaoying** [comment]
> it seems that after this cherry pick to 1.18, problem that "use of closed network connection" still re-produced. Should I also update go version ?

@liggitt could you help me on this issue? I have no idea what to do now

**Pull request merged**